bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1530 @@
+        impl_try_from_signed_type!($ty, i128);
+
+        impl fmt::Display for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
+        }
+
+        impl fmt::LowerHex for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::LowerHex::fmt(&self.0, f) }
+        }
+
+        impl fmt::UpperHex for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::UpperHex::fmt(&self.0, f) }
+        }
+
+        impl_div_mul!($ty, u8);
+        impl_div_mul!($ty, u16);
+        impl_div_mul!($ty, u32);
+        impl_div_mul!($ty, u64);
+
+        impl Shl<u32> for $ty {
+            type Output = Self;
+            #[inline]
+            fn shl(self, shift: u32) -> Self { Self(self.0 << shift) }
+        }
+
+        impl Shr<u32> for $ty {
+            type Output = Self;
+            #[inline]
+            fn shr(self, shift: u32) -> Self { Self(self.0 >> shift) }
+        }
+    };
+}
+
+/// A 256 bit integer representing work.
+///
+/// Work is a measure of how difficult it is to find a hash below a given [`Target`].
+///
+/// ref: <https://en.bitcoin.it/wiki/Work>
+///
+/// # Examples
+///
+/// ```
+/// use bitcoin::Work;
+/// let _ = Work::from(0xdeadbeef_u32);
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Work(U256);
+
+impl Work {
+    /// Lowest possible work value for Mainnet. See comment on [`Params::pow_limit`] for more info.
+    pub const MAINNET_MIN: Work = Work(U256(0x0000_0000_ffff_0000_0000_0000_0000_0000_u128, 0));
+
+    /// Lowest possible work value for Testnet. See comment on [`Params::pow_limit`] for more info.
+    pub const TESTNET_MIN: Work = Work(U256(0x0000_0000_ffff_0000_0000_0000_0000_0000_u128, 0));
+
+    /// Lowest possible work value for Signet. See comment on [`Params::pow_limit`] for more info.
+    pub const SIGNET_MIN: Work = Work(U256(0x0000_0377_ae00_0000_0000_0000_0000_0000_u128, 0));
+
+    /// Lowest possible work value for Regtest. See comment on [`Params::pow_limit`] for more info.
+    pub const REGTEST_MIN: Work = Work(U256(0x7fff_ff00_0000_0000_0000_0000_0000_0000_u128, 0));
+
+    /// Converts this [`Work`] to [`Target`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bitcoin::Work;
+    /// let w = Work::from(0xdeadbeef_u32);
+    /// assert_eq!(w.to_target().to_work(), w)
+    /// ```
+    pub fn to_target(self) -> Target { Target(self.0.inverse()) }
+}
+do_impl!(Work);
+
+impl Add for Work {
+    type Output = Work;
+    fn add(self, rhs: Self) -> Self { Work(self.0.wrapping_add(rhs.0)) }
+}
+
+impl Sub for Work {
+    type Output = Work;
+    fn sub(self, rhs: Self) -> Self { Work(self.0.wrapping_sub(rhs.0)) }
+}
```

Comment:
Shouldn't these panic with overflow checks enabled? I suggest to use `debug_assertions` for now and maybe we could do this insane trick later: https://users.rust-lang.org/t/detecting-overflow-checks/67698/8

Reply:
Oh I didn't know that overflow paniced in debug buids, TIL. I'll have to add back in `checked_add/sub/mul` to the `U256` type.